pkg/controllers/provisioning/scheduling/reservationmanager_test.go

```diff
@@ -0,0 +1,517 @@
+				rm.Reserve("hostname-3", testOffering1)
+
+				// Should have no capacity left
+				canReserve := rm.CanReserve("hostname-4", testOffering1)
+				Expect(canReserve).To(BeFalse())
+			})
+		})
+
+		Context("Multiple Reservations", func() {
+			It("should handle multiple offerings in a single call", func() {
+				rm.Reserve("hostname-1", testOffering1, testOffering2)
+
+				// Verify both reservations were made
+				canReserve1 := rm.CanReserve("hostname-1", testOffering1)
+				canReserve2 := rm.CanReserve("hostname-1", testOffering2)
+				Expect(canReserve1).To(BeTrue())
+				Expect(canReserve2).To(BeTrue())
+			})
+
+			It("should handle mixed new and existing reservations", func() {
+				// First reserve one offering
+				rm.Reserve("hostname-1", testOffering1)
+
+				// Then reserve both (one existing, one new)
+				rm.Reserve("hostname-1", testOffering1, testOffering2)
+
+				// Verify both are reserved and capacity is correctly tracked
+				canReserve1 := rm.CanReserve("hostname-1", testOffering1)
+				canReserve2 := rm.CanReserve("hostname-1", testOffering2)
+				Expect(canReserve1).To(BeTrue())
+				Expect(canReserve2).To(BeTrue())
+
+				// Verify capacity was only decremented once for testOffering1
+				canReserveOther1 := rm.CanReserve("hostname-2", testOffering1)
+				Expect(canReserveOther1).To(BeTrue()) // Should still have capacity
+			})
+		})
+
+		Context("Error Cases", func() {
+			It("should panic when trying to over-reserve", func() {
+				// Exhaust capacity
+				for i := 0; i < 3; i++ {
+					rm.Reserve(fmt.Sprintf("hostname-%d", i), testOffering1)
+				}
+
+				// Attempting to reserve more should panic
+				Expect(func() {
+					rm.Reserve("hostname-new", testOffering1)
+				}).To(Panic())
+			})
+
+			It("should panic with non-existent reservation ID", func() {
+				nonExistentOffering := &cloudprovider.Offering{
+					Available:           true,
+					ReservationCapacity: 1,
+					Requirements: pscheduling.NewLabelRequirements(map[string]string{
+						v1.CapacityTypeLabelKey:          v1.CapacityTypeReserved,
+						cloudprovider.ReservationIDLabel: "i-dont-exist",
+					}),
+				}
+				Expect(func() {
+					rm.Reserve("hostname-1", nonExistentOffering)
+				}).To(Panic())
+			})
+		})
+	})
+
+	Describe("Release", func() {
+		var testOffering1, testOffering2 *cloudprovider.Offering
+
+		BeforeEach(func() {
+			testOffering1 = offerings[0][0] // Use first offering from small-reserved (capacity 3)
+			testOffering2 = offerings[1][0] // Use first offering from medium-reserved (capacity 2)
+		})
+
+		Context("Valid Releases", func() {
+			It("should release a single reservation", func() {
+				// Reserve and then release
+				rm.Reserve("hostname-1", testOffering1)
+				rm.Release("hostname-1", testOffering1)
+
+				// Verify the reservation is no longer tracked for this hostname
+				// but capacity should be restored
+				canReserve := rm.CanReserve("hostname-2", testOffering1)
```

Comment:
I also don't see how this actually verifies that the release was successful - testOffering1 has a capacity of 3, even if we didn't release the reservation for hostname-1 we'd be able to make a reservation.